### ISSUE TITLE
Worker function parameter passing

### ISSUE THREAD

**Issue by alanz** [haskell/haskell-ide-engine#54]
A `CommandFunc` is defined as

```haskell
type CommandFunc = forall m. (MonadIO m,GHC.GhcMonad m,HasIdeState m)
                => [AcceptedContext] -> IdeRequest -> m IdeResponse
```

It will only be called by the dispatcher if at least one context matches, and and the required parameters are present and of the right type.

At the moment the command processing function is very messy

```haskell
checkCmd :: CommandFunc
checkCmd _ctxs req = do
  case getParams ["file"] req of
    Left err -> return err
    Right [ParamFile fileName] -> do
      liftIO $ runGhcModCommand (GM.checkSyntax [T.unpack fileName])
    Right x -> error $ "GhcModPlugin.checkCmd: got unexpected file param:" ++ show x
```

In this case there is only one valid context, so it it is called at all the "file" parameter is guaranteed to be valid. The `getParams` call returns a list of `ParamValue` in the order of the param names given to it, or a error result if they are missing. This call is actually redundant now.

It may be simpler to define the following types

```haskell
data ActualRequiredParams =
    AR0
  | AR1 ParamVal
  | AR2 ParamVal ParamVal
 ...
```
 
```haskell
data ActualOptionalParams =
    AO0
  | AO1 (Maybe ParamVal)
  | AO2 (Maybe ParamVal) (Maybe ParamVal)
 ...
```

as well as something similar for the matching contexts.

The `CommandFunc` would then be defined as

```haskell
type CommandFunc = forall m. (MonadIO m,GHC.GhcMonad m,HasIdeState m)
                => [ActualContextContext] -> ActualRequiredParams -> ActualOptionalParams -> m IdeResponse
```

This will still leave is with "missing case" warning messages though.

Is there some clever way to do this with types?

**Comment by alanz** [comment]
cc @JPMoresmau

**Comment by cocreature** [comment]
Should be possible to do via gadts (in fact I have just written a small prototype), but note that with 7.10 you'll still get warnings since the exhaustiveness checker is broken for GADTs, that's fixed in 8.0 (there was an icfp paper) afaik. IMHO it's still worth doing because then our implementation of getParams can't do stupid stuff, but since you seem to like simple code which I can totally understand (I just like playing around :)) I'd like to hear your opinion before cleaning up and making a PR.

**Comment by alanz** [comment]
I know we need to do something, because using convention-driven processes in haskell is silly.

Perhaps put something together, it has to be better than what we have now
